### ISSUE TITLE
Re-enable the Common & All tabs in the completion list (or otherwise improve the experience in certain cases)

### ISSUE THREAD

**Issue by dpoeschl** [dotnet/roslyn#4892]
(Port of internal bug 1213863)

EDIT: I originally thought the bug was about "Hide Advanced Members" option, but it's actually a request to bring back the Common & All tabs in the completion list. Some of the following conversation is about "Hide Advanced Members" and the ```EditorBrowsableAttribute``` -- those parts can be ignored, but are left for archiving purposes.

**Comment by dpoeschl** [comment]
Pinging VB folks: @AnthonyDGreen @ljw1004 (feel free to add others)

**Comment by AnthonyDGreen** [comment]
Is there a spec for this option? "Advanced" is too mysterious to illicit serious discussion. Without more details the option just sounds condescending "Why wouldn't I want the advanced members?". Unfortunately that response could just as well be leaving value on the table of it legitimately reduces irrelevant clutter.

**Comment by gchq** [comment]
Perhaps I was looking at the wrong menu item..

I was looking for the same option as in prior versions of VS where there are common members and all members, so that in response to a Boolean if I type Tr it will return True not Trace

Visual Studio 2010

![github001](https://cloud.githubusercontent.com/assets/3681474/9569680/9d2bc8a4-4f61-11e5-84d1-2eca6d36a64a.jpg)


Visual Studio 2013

![github002](https://cloud.githubusercontent.com/assets/3681474/9569684/af3a3602-4f61-11e5-81ec-68b1b6280529.jpg)

Visual Studio 2015

![github003](https://cloud.githubusercontent.com/assets/3681474/9569688/bac8cfa6-4f61-11e5-9c6f-ea272c460376.jpg)

**Comment by paul1956** [comment]
I agree with @gchq example, I really miss seeing only the common responses for auto competition  when using VS 2015. I don't remember every needing to use the "All" option. while using Vs 2013, maybe because if I want "Trace" and type "Tra" VS 2013 automatically switched to "All" without me noticing.

**Comment by Pilchie** [comment]
@AnthonyDGreen `Advanced` means marked with `<EditorBrowsable(EditorBrowsableState.Advanced)>`.

This is *not* the population of the Common/All tabs, which is done based on XML doc comment tags.

**Comment by gchq** [comment]
..and we can edit this to show Common/All how?

**Comment by Pilchie** [comment]
Roslyn currently has no support for Common/All.  I meant in the old compiler.

**Comment by gchq** [comment]
When you say currently no support, does this mean it will be at some stage? 
When you are steaming through code it makes a hell of a difference, especially when you don't want to break your train of thought. It's like juggling with 20 balls and when they all come crashing down you have to figure out where you were heading all over again...........

**Comment by dpoeschl** [comment]
Sorry, I had misread @gchq's original bug report. The request is to support the Common/All tabs again, not to make "Hide Advanced Members" work. This leaves no request for VB "Hide Advanced Members", so we can drop that discussion. I've updated the title of this bug to reflect the actual request from @gchq.

**Comment by Pilchie** [comment]
Okay - given the user feedback, I think we need to do something to support Booleans/Enums/CompletionSet.  We need to come up with the right design though.

**Comment by rchande** [comment]
Sounds like the problem is that there are cases where preselection targets a category of things (Booleans, enum members, "completionlist"), but they aren't always grouped together in the list. Enum members would be, but booleans aren't and "completionlist" things might not be.

Instead of generic "Common" and "All" tabs that are present all the time, I wonder if we could add tabs to the list only when preselection is specializing the list to a category of things. Furthermore, to indicate that one tab was for the specialized list of things, I would name the tab after the category of things being displayed. See my MSPaint mockups below. Not sure what we'd call the other tab, but it would include the full list that we already show today. For "new" preselection, since there's only one choice to display, we would show the full list. Finally, typing something with no matches in the specialized tab would automatically switch you over to the other one.

Not sure how we would handle persisting the choice of tab--maybe users would always want to start in the specialized tab if it was available?

![list](https://cloud.githubusercontent.com/assets/3751401/9611930/8c914e9c-5097-11e5-8dfe-2bd0f217b6c6.png)

Thoughts @AnthonyDGreen @DustinCampbell @Pilchie?

**Comment by paul1956** [comment]
I don't understand the requirement for a new design Visual Studio through 2013 know how to display the stuff I needed 99.999% of the time with me no doing nothing, if what I needed wasn't shown it was obvious that I could select "all" and get the rest. For booleans true was displayed when I hit T and nothing else. For enums I got the various values sorted alphabetically. What would be the usage case where I want to see these huge lists all the time?

**Comment by DustinCampbell** [comment]
@paul1956, historically the Common/All tabs have been problematic for VB. The APIs that are included in the Common tab have not been curated for several releases (probably not since I did it for Visual Studio 2010). So, there would likely be many "common" APIs missing from the Common tab if we fully re-enabled it. There were several other problems as well, here are a few that we'd like to avoid.

* Once a developer selects the All tab, the Common tab is no longer shown by default. And, because this is persisted across Visual Studio sessions the developer will see the All tab by default when they start Visual Studio the next time.
* Extension methods were only available in the All tab, which became problematic as more extension methods became common APIs.
* The APIs that are placed in the Common tab were curated. In other words, they were hand-picked and this was a very error-prone process. And, as I mentioned, it hasn't happened in several years. :disappointed: 

Instead of adding Common/All tabs back everywhere and dealing with the same old problems, we'd like to focus on the real pain points that were introduced with their removal. From lots of customer feedback (including your screenshots above), we believe there are three specific cases that need to be addressed:

1. Boolean values for True and False
2. Enum values
3. Anything with the <completionset/> XML doc comment. This is the system that underpins the special completion lists that appear for specific types such as "System.Drawing.Color".

We believe that this should cover the majority of scenarios that customers are having trouble with. What do you think?

**Comment by gchq** [comment]
Anything that reduces the amount going into the swear jar, or out of the brandy bottle, is a step in the right direction for me :-)

**Comment by paul1956** [comment]
@DustinCampbell first I want to thank you for what you implemented in 2010, I never noticed the issues you list above which says for many developers it worked as desired. I agree your list is the three places I used the feature the most, the third seems like the least used but most difficult to do without. As long as the following example works like in 2013 where I am offered TrueBlue, when I type Trueb, I am fine with your proposal.
```
        Dim x As Boolean
        Dim TrueBlue As Boolean = False
        If x = TrueBlue Then

        End If
```

**Comment by ljw1004** [comment]
I really like the discussion in this thread. It's difficult for me to put into words, because I have daily first-hand frustration at the completion list (overwhelming and full of junk and somehow sticks in "ArgIterator" into my code too often). And yet I don't know how to conceptualize what I want. Dustin your breakdown is helpful.

But it still doesn't quite feel right. For instance, I think I never write `Dim x As Boolean = True|False` in my code -- there's rarely any point! For a boolean, I'm more likely to have the expression on the right be another local variable or a function call or a relational operator.

![boolean](https://cloud.githubusercontent.com/assets/3316258/9619013/372575e0-50c0-11e5-8251-84fbe14f1452.png)

**The full list for Boolean is about 1000 members long!** (I tried to count, got up to 230, saw I was about a quarter of the way through, and gave up). This is in a new console app without any Imports statements. Once I add them in, there'll be more. Let's work through just the ones pictured...

 * `Dim x As Boolean = Base64FormattingOptions|` -- *never*
 * `Dim x As Boolean = Beep|` - *never*
 * `Dim x As Boolean = BitArray|` - *never*, since the only useful members on BitArray are instance members
 * `Dim x As Boolean = BitConverter.ToBoolean(buf,0)` - *yes*, this is useful
 * `Dim x As Boolean = Boolean.Parse("true")` - *yes*, this is useful
 * `Dim x As Boolean = BooleanSwitch|` - *never*, since the only useful members on BooleanSwitch are instance members
 * `Dim x As Boolean = Buffer|` - *no*. There are a few conceivable shared members like `Buffer.GetByteLength(buf) > 10` but that's pretty out-there
 * `Dim x As Boolean = Byte.TryParse(...)` - *yes*, this is useful
 * `Dim x As Boolean = CallByName(..)` - *no*. There are a few conceivable late-bound option-strict-off scenarios, but they're pretty weird.


Here's some criteria I'll use to assess how often something is used based on *context*, where context refers to the syntactical position (e.g. parameter argument, subexpression, Dim expression) and to the expected target type.
 1. Things that are used >30% of cases in this context
 2. Things that are used >0.5% of cases in this context
 3. Things that are used <0.5%
 4. Things that are *never* used

I suspect that `True|False` probably meets criteria [1] for booleans in parameter arguments and subexpressions, and meets criteria [2] for booleans in Dim statements. I suspect that local variables meet criteria [1] for Booleans in all positions.


Where my post going? I don't know. I don't have a concrete proposal. But if you merely add a tab for Booleans and enums, and retain "All" which contains everything else, (1) I think the Boolean tab will be too small a default for lots of Boolean cases, (2) I think the All tab will still be too huge, including as it does all those useless things.


Currently, my only defense against the huge 1000+ list is Hungarian notation - I name my related local variables or my member variables with a common prefix, e.g. "queueSize, queueWeight, queueHead, queueTail", so that by typing the first few letters I'll land in the right place in the completion list.

**Comment by DustinCampbell** [comment]
Thanks a ton for the feedback Lucian. We have ideas for target typing completion lists but, as you know, it's a particularly challenging problem. That said, you've identified at least one category of items that we could just filter out now: global Subs. (There's no reason that Beep() should be in there. :smile:)

In this case, I think we need to address the initial problem of no longer being about to type ```Dim x As Boolean = t<TAB>```. That's causing a fair amount of muscle memory pain. However, I'm interested in your ideas for a more general solution. Maybe you, @Pilchie, @rchande and I can chat at some point?

**Comment by Alirezanet** [comment]
bump . it was great future . bring back it pls :(

**Comment by rchande** [comment]
We've done some work to preselect the "True" and "False" keywords. I hope people can try it out in update 1 and see if that helps the experience match what they were used to.

**Comment by Pilchie** [comment]
For the record this work was in #6360.  Moving this issue to 1.2, where we will see how things are with the improved preselection and decide if we want to do more.

**Comment by rchande** [comment]
Hi @paul1956 @gchq @Alirezanet:
I merged #6318 for Update 1. Has that helped with this scenario for you?

**Comment by mairaw** [comment]
Our current CMS is not adding the FilterPriority tag anymore for the new IntelliSense builds and I'm trying to determine if that's a bug or not. Do we still need that in our .NET IntelliSense files? Thanks.

**Comment by rchande** [comment]
We don't currently check it and I don't think we plan ever to do so. Can you confirm @Pilchie @DustinCampbell ?

**Comment by AnthonyDGreen** [comment]
Why is it no longer adding the tag?

**Comment by mairaw** [comment]
It would be great to get confirmation. We changed to a new system last year and the team developing were probably not aware of that attribute. I just noticed this recently when running some tests and I found this thread today while trying to learn more about the effects of that.

**Comment by AnthonyDGreen** [comment]
At this time we're trying to keep the coding experience in VB and C# mostly aligned and are being very intentional about when we pursue differences. Customer feedback and usage is one factor that contributes to that intention. The Common & All tabs which take advantage of this data are such a difference which has existed historically but which we did not include in the RTM release of Visual Studio 2015. Since then we've received some feedback regarding that decision from the VB community and have been working with them to identify the highest value scenarios to target in an attempt to restore the most familiarity to their development experience while, again, keep the overall experiences between the languages consistent.

That said, we're still open to additional feedback in this area. Therefore I would say it would be most prudent to ensure the tool passes through the same data it's passed through before, for the sake of older versions of Visual Studio targeting newer libraries, any third party tools which may depend on this data, and to ensure your libraries are resilient to whatever decisions (or indecision) we make in this area in the future. Given where we are today I wouldn't say treat this with priority-0 urgency but I don't think it wise to drop any data in anticipation that we don't make any change in this area in the future.

**Comment by mairaw** [comment]
Sounds good @AnthonyDGreen. I'll pursue the fix but won't consider blocking. Thanks!

**Comment by Pilchie** [comment]
We have not seen any of the same feedback since Update 1 was released with #6318.  It seems to have been sufficient.
